### PR TITLE
refactor(trimming): implement pre-trimming

### PR DESCRIPTION
**Pull request by niklasad1**
This commit refactors trimming such that the staking-miner is responsible to ensure that trimming doesn't have to be done be on-chain.

The pre-trimming is implemented as follows:

1. Weight trimming: Initially the voters are iteratively removed until Weight < MaxWeight by using MinerConfig::solution_weight for this part a "new solution" doesn't need to be "mined" and is much more efficient than part 2 and only has to be performed once.
2. Length Trimming: Instead of removing one voter at the time and then “mine a new solution” as the initial implementation did. Now the implementation removes the number of voters in a binary search manner to find "the biggest set of voters" where no trimming occurs such that it doesn't get trimmed.

By “binary search manner” I mean:

```rust
let  mut  l  =  1;
let  mut  h  =  voters.len();
let  mut  best_solution  =  None;

while  l  <=  h {
   let  mid  = ((h  -  l) /  2) +  l;
   let next_state = voters.trim(mid)?;
   let solution = mine_solution(next_state);
	 
   if !solution.is_trimmed() {
     best_solution  =  Some((solution, score, solution_or_snapshot_size));
     h  =  mid  -  1;
   } else {
     l  =  mid  +  1;
   }
}
```

This makes things much faster than the iterative approach I implemented first

This PR also adds a prometheous metric to count the number of "pre-trimming length ops"

### CODE REVIEW

File: staking-miner-playground/runtime/src/lib.rs
Comment (niklasad1) [inline staking-miner-playground/runtime/src/lib.rs:335]: Because the trimming is slower with "pre-trimming", I increased the SignPhase a bit...

File: staking-miner-playground/runtime/src/lib.rs
Comment (niklasad1) [inline staking-miner-playground/runtime/src/lib.rs:461]: modified these values to ensure trimming always occurs when we run it.

File: Cargo.toml
Comment (niklasad1) [inline Cargo.toml:27]: change to master once we have merged the substrate PR

File: src/epm.rs
Comment (gpestana) [inline src/epm.rs:71]:
```suggestion
/// The trimming works by removing the voter with the least amount of stake.
```

File: src/epm.rs
Comment (gpestana) [inline src/epm.rs:94]: Have we tried to trim more than one voter in each iteration? This may reduce the number of iterations required to finalise the trimming. Maybe we could have an heuristic where depending on how much the solution exceeds the limit, trim 1 or more voters in one go.

This probably should not block this PR, but perhaps addressed in a subsequent PR if helpful.

File: src/epm.rs
Comment (gpestana) [inline src/epm.rs:325]: As per my comment above, perhaps we could pass the number of voters to trim which will depend on the iteration we are at. E.g. if this is the first call to `fn trim_next` in the loop, trim N voters. After a few attempts, start to trim only one voter since we're approaching the solution that does not need any more trimming.

File: src/epm.rs
Comment (gpestana) [reply src/epm.rs:94]: We could also calculate the difference between the `weight_trimmed`/`length_trimmed` calculated in  `prepare_election_result_with_snapshot` and use that as an heuristic of how much trimming will be required. Then we can use that to request the trimming of N voters in one go, if there's still a long way to trim.

**Review by gpestana — APPROVED** [review]
Added a few comments that could optimise the number of trimming round before reaching the appropriate solution weight/length, but I think that should not be a blocker for this PR if it's too much work.

Thanks for implementing this!

File: src/epm.rs
Comment (niklasad1) [reply src/epm.rs:94]: That's good point, I think we should merge this PR and create a issue for your idea.

If there's much trimming to do then yeah this would be very slow as it does it one-by-one.

File: src/epm.rs
Comment (niklasad1) [inline src/epm.rs:136]: decided to add this in the constructor/new because it should only be needed to do once

File: src/epm.rs
Comment (niklasad1) [inline src/epm.rs:402]: binary search to find the best "pre-trim"

**Pull request merged**
